### PR TITLE
Add support for only closing container once per X time period due to SweepReady Object Used

### PR DESCRIPTION
**Pull request by markfields**
This is a follow-on to #11963.

To help us and our partners flush out bugs with how datastores are referenced in Dogfood environments, #11963 added an option which will close the container when an object is used after being marked as "SweepReady". This means the Sweep Timeout has elapsed since it was marked as unreferenced.

Closing the container at this time will cause some user pain - users will be locked out of their documents. So this change adds a mechanism using localStorage by which we can set a "blackout period" such that a given document will only close once per <configurable time period> with this error.

This will allow us to balance the user pain v. usefulness from diagnostic standpoint to enable our partners and us to track down the remaining sweep-blocking bugs - Until all such bugs are fixed, turning on Sweep represents a Data Corruption liability.

### CODE REVIEW

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            || Date.now() > lastCloseTime + blackoutPeriodDays * oneDayMs; // ...we've passed the blackout period
+
+        if (shouldClose) {
+            // Update closures map in localStorage before closing
+            pastClosuresMap[this.uniqueContainerKey] = { lastCloseTime: Date.now() };
+            this.localStorage.setItem(closuresStorageKey, JSON.stringify(pastClosuresMap));
+
+            this.closeFn(error);
+        } else {
+            this.mc.logger.sendErrorEvent({ eventName: "IgnoringSweepReadyObjectUsage" }, error);
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:127]: I would prefer SweepReadyObjectUsage

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        const blackoutPeriodDays = this.mc.config.getNumber(blackoutPeriodDaysKey);
+
+        const error = new SweepReadyUsageError(
+            "SweepReady object used in Non-Summarizer Client",
+            { errorDetails: JSON.stringify({ ...errorProps, lastCloseTime, blackoutPeriodDays }) },
+        );
+
+        // Should close if...
+        const shouldClose = lastCloseTime === undefined // ...We've not closed before, or...
+            || blackoutPeriodDays === undefined // ...there's no blackout duration set, or...
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:117]: What's blackoutPeriodDays?

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+ */
+export class SweepReadyUsageDetectionHandler {
+    private readonly localStorage: Pick<Storage, "getItem" | "setItem">;
+
+    constructor(
+        /** Unique key for this container for diagnostic purposes */
+        private readonly uniqueContainerKey: string,
+        private readonly mc: MonitoringContext,
+        private readonly closeFn: (error?: ICriticalContainerError) => void,
+        localStorageImpl?: Pick<Storage, "getItem" | "setItem">,
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:64]: Why aren't we setting this in the main code path?

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts [hunk 4]
```diff
@@ -0,0 +1,132 @@
+/*!
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:1]: Not part of this PR - We should move GC files into its own folder. Then we can break more functionality out of the main GC file. There are parts that should be in separate file but I did not do it because I didn't want too many files under container-runtime/src folder.

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts (hunk 1 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:129]: Well we already have the other logs.  Although those only go on Loaded - maybe we should update that to always log even for main container, and it's up to us to filter out to get better results.

@agarwal-navin thoughts?

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    /**
+      * If SweepReady Usage Detection is enabled, close the main container.
+      * If the "Blackout Period" is set, don't close the container more than once in that period.
+      *
+      * Once Sweep is fully implemented, this will be removed since the objects will be gone
+      * and errors will arise elsewhere in the runtime
+     */
+    public usageDetectedInMainContainer(errorProps: ITelemetryProperties) {
```
[the remark this answers is not in the document]
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:90]: It's in the comments.  Can't think of a better name. I can add a comment to the call site too.  Lmk if you have a better name, happy to change it.

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts (hunk 2 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:117]: A little more details on how this will be used will be helpful :-)

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts (hunk 4 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:1]: Yeah I took a middle-ground of staying in the same folder but putting gc prefix - they'll all be adjacent in the directory.  Hesitating on doing the more disruptive directory while we're at a critical time in rolling this out - if we want to be patching between releases it'll make it more painful (even moving code makes it a little painful, but felt worth it).

There are some things to consider when adding sub-dirs, I made a post about it in the Dev channel while you were OOF. Don't remember what they are.  Maybe we open a backlog item for it, it will take some effort.

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts (hunk 3 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:63]: We'll use browser's localStorage (see 71: `this.localStorage = localStorage ?? noopStorage`).

I could call it override instead of impl

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        // Should close if...
+        const shouldClose = lastCloseTime === undefined // ...We've not closed before, or...
+            || blackoutPeriodDays === undefined // ...there's no blackout duration set, or...
+            || Date.now() > lastCloseTime + blackoutPeriodDays * oneDayMs; // ...we've passed the blackout period
+
+        if (shouldClose) {
+            // Update closures map in localStorage before closing
+            // Note there is a race condition between different tabs updating localStorage and overwriting
+            // each others' updates. If so, some tab will crash again. Just reload one at a time to get unstuck
+            pastClosuresMap[this.uniqueContainerKey] = { lastCloseTime: Date.now() };
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:130]: Is this key supposed to be unique per document or per container? Currently, you are using `ContainerContext::id` which is the document id.

Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:130]: Per document.  Needs to prevent the next session from closing.

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts (hunk 2 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:123]: Take a look at latest iteration.  Also this function's doc comments mentions the term.  Where would be the best place to explain how this works?

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+export class SweepReadyUsageError extends LoggingError implements IFluidErrorBase {
+    /** This errorType will be in temporary use (until Sweep is fully implemented) so don't add to any errorType type */
+    public errorType: string = "unreferencedObjectUsedAfterGarbageCollected";
+}
+
+/**
+ * This class encapsulates the logic around what to do when a SweepReady object is used.
+ * There are several tactics we plan to use in Dogfood environments to aid diagnosis of these cases:
+ *  - Closing the main container when either the main or summarizer client detects this kind of violation
+ *  - Throttling the frequency of these crashes via a "Blackout Period" per container per device
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:60]: Can you also add comment here or in near the implementation how different settings affect this.
For example, a setting controls whether this whole functionality of closing container is enabled. A second setting controls whether we want to close it all the time or once in a period of time. Which is the second setting - blackoutPeriodDaysKey?

Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:60]: Good idea to mention settings here, thanks

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (tyler-cai-microsoft) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:96]: `errorOnSweepReadyUsageInMainContainer`?

[the remark this answers is not in the document]
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:96]: But it might not, if the setting is disabled 🤔

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+      *
+      * Once Sweep is fully implemented, this will be removed since the objects will be gone
+      * and errors will arise elsewhere in the runtime
+     */
+    public usageDetectedInMainContainer(errorProps: ITelemetryProperties) {
+        if (!sweepReadyUsageDetectionSetting.read(this.mc.config).mainContainer) {
+            return;
+        }
+
+        const pastClosuresMap: Record<string, { lastCloseTime: number; } | undefined> = (() => {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:101]: I am wondering if you could avoid reading the local storage and set up this state if `blackoutPeriodDaysKey` is undefined and just close the container. If its not, do all of this logic. It will make the code more readable and easier to understand :-).

Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:103]: I thought of that but thought the code was more readable this way :P  I'll take a quick look again.

Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:103]: Yeah this will work out and read better, you're right - thanks!

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts (hunk 2 above)
Comment (tyler-cai-microsoft) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:125]: If I understand correctly, blackoutPeriodDays is the number of days this error gets ignored if it's not the first time it occurs.

Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:125]: Yep.  Will be clearer after this little refactor I think.

File: packages/runtime/container-runtime/src/garbageCollection.ts [hunk 9]
```diff
@@ -1457,16 +1433,8 @@ export class GarbageCollector implements IGarbageCollector {
...
-            // and errors will arise elsewhere in the runtime
-            if (state === UnreferencedState.SweepReady &&
-                sweepReadyUsageDetectionSetting.read(this.mc.config).mainContainer
-            ) {
-                const error = new SweepReadyUsageError(
-                    "SweepReady object used in Non-Summarizer Client",
-                    { errorDetails: JSON.stringify(propsToLog) });
-                this.runtime.closeFn(error);
+            if (state === UnreferencedState.SweepReady) {
+                this.sweepReadyUsageHandler.usageDetectedInMainContainer(propsToLog);
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/garbageCollection.ts:1440]: You should add `usageType` here so we know whether the object was loaded, revived or used.

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts (hunk 6 above)
Comment (tyler-cai-microsoft) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:132]: Basically if 10 different clients open the document, they will all close once if I understand this correctly.

Comment (tyler-cai-microsoft) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:132]: This bug will be very hard to reproduce on the consumer end

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts (hunk 1 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:137]: I think this is fine for now. As a follow up, we should probably move all the inactive / sweep ready logging logic here and then we can remove duplicates.
Maybe `SweepReadyObject_Used` to make the part after _ unique?

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts [hunk 10]
```diff
@@ -964,37 +903,108 @@ describe("Garbage Collection Tests", () => {
             );
         });
 
-        describe("SweepReady usage error enabled (main container)", () => {
-            const snapshotCacheExpiryMs = 500;
-            const sweepTimeoutMs = defaultSessionExpiryDurationMs + snapshotCacheExpiryMs + oneDayMs;
+        describe("Main Container Behavior", () => {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:906]: Not sure how I feel about the term "main container". I feel like this is adding yet another term to the ones already there :-). For instance, in the summarizer world, the terms are "summarizer client" and "parent client" that spawns the "summarizer client". All other clients are just "regular clients".

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+ *  - Throttling the frequency of these crashes via a "Blackout Period" per container per device
+ */
+export class SweepReadyUsageDetectionHandler {
+    private readonly localStorage: Pick<Storage, "getItem" | "setItem">;
+
+    constructor(
+        private readonly uniqueContainerKey: string,
+        private readonly mc: MonitoringContext,
+        private readonly closeFn: (error?: ICriticalContainerError) => void,
+        localStorageImpl?: Pick<Storage, "getItem" | "setItem">,
```
[the remark this answers is not in the document]
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts]: In tests, since localStorage doesn't exist in node.

File: packages/runtime/container-runtime/src/garbageCollection.ts (hunk 9 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/garbageCollection.ts:1440]: Good catch

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts (hunk 1 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:141]: Ok sure I can match the format of the other logs, good idea.

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+import { ICriticalContainerError } from "@fluidframework/container-definitions";
+import {
+    IConfigProvider,
+    IFluidErrorBase,
+    LoggingError,
+    MonitoringContext,
+} from "@fluidframework/telemetry-utils";
+import { oneDayMs } from "./garbageCollection";
+
+/** Feature Gate Key -
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:16]: nit: comment should start at next line.

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:98]: I've added comments on the class, function, and call site.  Leaving the name as-is.

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts (hunk 10 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:906]: Oh ok good to know.  I can have a follow-up PR to clean up the naming.  "Parent Client" is better?

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts (hunk 12 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:16]: I thought something was off!  Thanks

**Comment by agarwal-navin** [comment]
I am trying to understand how this will be used in dog food. If a user reports that their container keeps on closing and they can't make progress, we ask them to set the localStorage flag for black out days?

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts (hunk 6 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:139]: Tyler - You are right that if we go this route, we'll lose a lot of leverage with users in terms of getting good bug reports.  When I first reach out to partners, I'm not planning to mention this capability.  But it's there, so if the conversation starts to go towards "We have to disable SweepReady Usage Detection, the user impact is too high", then we will bring this out.

In other words, it gives us a dial on user pain so we can tune it to the org's tolerance for 💩 user experience in DF

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+ *  How many days between closing the container from this error (avoids locking user out of their file altogether)
+ */
+export const blackoutPeriodDaysKey = "Fluid.GarbageCollection.Dogfood.SweepReadyUsageDetection.BlackoutPeriodDays";
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:20]: SkipClosureForDays? Just to make it very explicit what it means? Blackout period kind of sounds the opposite right, like when airlines have blackout days, you don't get any discounts, etc. :D

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts (hunk 6 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:139]: Also, not sure you said it quite right - the pastClosuresMap is in localStorage in that browser (or Teams or whatever).  So every user gets the crash once per "browser" per X days.  Still pretty annoying IMO.

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts (hunk 13 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:20]: haha ok yeah sure, thanks.

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts (hunk 6 above)
Comment (tyler-cai-microsoft) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:139]: True, I did not try to get that specific

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    LoggingError,
+    MonitoringContext,
+} from "@fluidframework/telemetry-utils";
+import { oneDayMs } from "./garbageCollection";
+
+/**
+ * Feature Gate Key -
+ * How many days between closing the container from this error (avoids locking user out of their file altogether)
+ */
+export const blackoutPeriodDaysKey = "Fluid.GarbageCollection.Dogfood.SweepReadyUsageDetection.SkipClosureForXDays";
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:20]: nit: Rename the variable too.

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+        }
+
+        // Default stance is we close every time - this reflects the severity of SweepReady Object Usage.
+        // However, we may choose to "throttle" the closures by setting the BlackoutPeriodDays setting,
+        // which will only allow the container to close once during that period, to avoid locking users out.
+        let shouldClose: boolean = true;
+        let pastClosuresMap: Record<string, { lastCloseTime: number; } | undefined> = {};
+        let lastCloseTime: number | undefined;
+        const blackoutPeriodDays = this.mc.config.getNumber(blackoutPeriodDaysKey);
+        if (blackoutPeriodDays !== undefined) {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:113]: Should you skip this if localStorage is noopStorage?

**Review by agarwal-navin — APPROVED** [review]

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts (hunk 15 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:113]: It works as-is (there's test coverage), that's an unexpected case, so not wanting to call it out.

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts (hunk 14 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:20]: I decided not to, that term is all over the place in variable names and comments.  I'll do a quick follow-up PR though.

**Comment by markfields** [comment]
> I am trying to understand how this will be used in dog food. If a user reports that their container keeps on closing and they can't make progress, we ask them to set the localStorage flag for black out days?

As discussed in Teams - Not expecting to hand this out to users, but rather work with our partners if they push back on DF experience impact, to set it via ECS.

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts (hunk 10 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts]: Renamed to interactive client

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts (hunk 14 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts]: I did it here after all

File: api-report/container-runtime.api.md [hunk 16]
```diff
@@ -346,7 +346,7 @@ export interface IEnqueueSummarizeOptions extends IOnDemandSummarizeOptions {
 
 // @public
 export interface IGarbageCollectionRuntime {
-    closeFn(error?: ICriticalContainerError): void;
+    closeFn: (error?: ICriticalContainerError) => void;
```
Comment (markfields) [inline api-report/container-runtime.api.md:349]: @agarwal-navin why is this exported?

Comment (markfields) [reply api-report/container-runtime.api.md:349]: I mean why is `IGarbageCollectionRuntime` exported.  (I had to change how closeFn was written to work around an incorrect lint error detection)
